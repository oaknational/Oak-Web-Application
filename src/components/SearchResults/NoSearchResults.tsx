import Flex from "../Flex";
import { Heading, P, Span } from "../Typography";

interface NoSearchResultsProps {
  searchTerm: string;
}

const NoSearchResults = (props: NoSearchResultsProps) => {
  const { searchTerm } = props;

  return (
    <Flex $flexDirection="column" $pl={24}>
      <Heading tag={"h4"} $mt={24} $mb={16} $font={"heading-7"}>
        No search results
      </Heading>
      <Flex $flexDirection="column">
        <P $font={"body-1"} $mb={12}>
          Sorry, we could not find any results for{" "}
          <Span $wordWrap={"break-word"}>“{searchTerm}”</Span>.
          <br /> <br /> Please enter a topic you wish to explore in the search
          bar above.
        </P>
      </Flex>
    </Flex>
  );
};

export default NoSearchResults;
