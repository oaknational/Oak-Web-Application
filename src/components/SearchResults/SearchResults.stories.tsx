import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { searchResultsHitsSchema } from "../../context/Search/helpers";
import elasticResponseFixture from "../../context/Search/elasticResponse.fixture.json";

import Component from "./SearchResults";

export default {
  title: "Lists/Search Results",
  component: Component,
  argTypes: {},
} as ComponentMeta<typeof Component>;

const Template: ComponentStory<typeof Component> = (args) => (
  <Component {...args} />
);

const hits = searchResultsHitsSchema.parse(elasticResponseFixture.hits.hits);

export const SearchResults = Template.bind({});
SearchResults.args = {
  hits,
};
