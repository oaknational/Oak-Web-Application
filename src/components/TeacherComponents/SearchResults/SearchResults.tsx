import {
  OakLI,
  OakUL,
  OakFlex,
  OakPagination,
  OakBox,
} from "@oaknational/oak-components";

import SignPostToAila from "../NoSearchResults/SignPostToAila";

import usePagination from "@/components/SharedComponents/Pagination/usePagination";
import {
  KeyStage,
  SearchHit,
  SearchQuery,
} from "@/context/Search/search.types";
import SearchResultsItem, {
  SearchResultsItemProps,
} from "@/components/TeacherComponents/SearchResultsItem";
import { getSearchHitObject } from "@/context/Search/search.helpers";

interface SearchResultsProps {
  hits: Array<SearchHit>;
  allKeyStages: KeyStage[];
  searchResultOpened: (
    searchHit: SearchResultsItemProps,
    searchRank: number,
  ) => void;
  searchResultExpanded: (
    searchHit: SearchResultsItemProps,
    searchRank: number,
  ) => void;
  query?: SearchQuery;
}

export const RESULTS_PER_PAGE = 20;

const SearchResults = (props: SearchResultsProps) => {
  const {
    query,
    hits,
    allKeyStages,
    searchResultOpened,
    searchResultExpanded,
  } = props;
  const hitCount = hits.length;
  const paginationProps = usePagination({
    totalResults: hitCount,
    pageSize: RESULTS_PER_PAGE,
    items: hits,
  });

  const { currentPageItems, currentPage, firstItemRef, paginationRoute } =
    paginationProps;
  const currentPageItemsWithSignPost = [
    ...currentPageItems.slice(0, 3),
    "aila-sign-post",
    ...currentPageItems.slice(3),
  ];
  const searchRank = (index: number) => {
    return (currentPage - 1) * 20 + index + 1;
  };

  return (
    <OakFlex $background={"white"} $flexDirection="column" id="search-results">
      {hitCount ? (
        <OakUL $reset>
          {currentPageItemsWithSignPost.map((hit, index) => {
            if (typeof hit !== "string") {
              const searchHitObject = getSearchHitObject(hit, allKeyStages);
              if (!searchHitObject) {
                return null;
              }
              return (
                <OakLI
                  key={`SearchList-SearchListItem-${index}${hit._source.slug}`}
                >
                  <SearchResultsItem
                    {...searchHitObject}
                    firstItemRef={index === 0 ? firstItemRef : null} // this is for pagination focus
                    onClick={(props) => {
                      searchResultOpened(props, searchRank(index));
                    }}
                    onToggleClick={(props) =>
                      searchResultExpanded(props, searchRank(index))
                    }
                  />
                </OakLI>
              );
            }
            return (
              <SignPostToAila
                title="Can't find what you need?"
                text="Create a tailor-made lesson plan and resources on any topic with Alia, our free AI-powered lesson assistant. Entirely adaptable to your class and context."
                searchExpression={query?.term}
                keyStage={
                  query?.keyStages?.length === 1 ? query.keyStages[0] : ""
                }
                subject={query?.subjects?.length === 1 ? query.subjects[0] : ""}
              />
            );
          })}
        </OakUL>
      ) : null}

      {hits.length > RESULTS_PER_PAGE && (
        <OakBox
          $width="100%"
          $mt={["space-between-none", "auto"]}
          $pb="inner-padding-xl7"
          $pt="inner-padding-xl4"
        >
          <OakPagination
            {...paginationProps}
            pageName={"Search"}
            paginationHref={paginationRoute}
          />
        </OakBox>
      )}
    </OakFlex>
  );
};

export default SearchResults;
