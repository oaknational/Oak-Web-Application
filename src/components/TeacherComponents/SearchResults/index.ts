export { default } from "./SearchResults";
export { RESULTS_PER_PAGE } from "./SearchResults";
