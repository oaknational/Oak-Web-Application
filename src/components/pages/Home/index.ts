export { default as SharedHomeContent } from "./SharedHomeContent";
export { default as HomeSiteCards } from "./HomeSiteCards";
