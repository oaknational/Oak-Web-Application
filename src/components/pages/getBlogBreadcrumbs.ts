import { BlogWebinarCategory } from "../../common-lib/cms-types";
import { SerializedBlog } from "../../pages/blog/[blogSlug]";
import { Breadcrumb } from "../Breadcrumbs";

/*
 *
 * getBlogBreadcrumbs()
 *
 * takes a list of blog categories and the slug of the selected category
 * returns an array of breadcrumbs - blog > [category]
 *
 */

export const getBlogBreadcrumbs = (
  categories: BlogWebinarCategory[],
  currentCategorySlug: string | null
): Breadcrumb[] => [
  { label: "Blog", href: "/blog" },
  {
    label:
      categories.find((cat) => cat.slug === currentCategorySlug)?.title ||
      "All",
    href: currentCategorySlug || "/blog",
    disabled: true,
  },
];

export const getBlogPostBreadcrumbs = (
  categories: BlogWebinarCategory[],
  blog: SerializedBlog
): Breadcrumb[] => {
  const { title, slug, category } = blog;
  return [
    { label: "Blog", href: "/blog" },
    {
      label:
        categories.find((cat) => cat.slug === category.slug)?.title || "All",
      href: `/blog/categories/${category.slug}`,
    },
    {
      label: title,
      href: slug,
      disabled: true,
    },
  ];
};
