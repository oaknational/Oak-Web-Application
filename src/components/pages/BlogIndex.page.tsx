import { uniqBy } from "lodash/fp";
import {
  GetStaticPaths,
  GetStaticProps,
  GetStaticPropsResult,
  NextPage,
} from "next";
import { useTheme } from "styled-components";

import { BlogListJsonLd } from "../../browser-lib/seo/getJsonLd";
import { getSeoProps } from "../../browser-lib/seo/getSeoProps";
import CMSClient from "../../node-lib/cms";
import {
  BlogPostPreview,
  BlogWebinarCategory,
} from "../../common-lib/cms-types";
import { decorateWithIsr } from "../../node-lib/isr";
import BlogCategoryList from "../Blog/BlogCategoryList";
import useBlogCategoryList from "../Blog/BlogCategoryList/useBlogCategoryList";
import BlogList from "../Blog/BlogList";
import { BlogListItemProps } from "../Blog/BlogList/BlogListItem";
import Box from "../Box";
import SummaryCard from "../Card/SummaryCard";
import Grid, { GridArea } from "../Grid";
import Layout from "../Layout";
import MaxWidth from "../MaxWidth/MaxWidth";
import MobileBlogFilters from "../MobileBlogFilters";
import { Heading } from "../Typography";

import { getBlogBreadcrumbs } from "./getBlogBreadcrumbs";

export type SerializedBlogPostPreview = Omit<BlogPostPreview, "date"> & {
  date: string;
};

export type BlogListingPageProps = {
  blogs: SerializedBlogPostPreview[];
  categories: BlogWebinarCategory[];
  categorySlug: string | null;
};

const BlogListingPage: NextPage<BlogListingPageProps> = (props) => {
  const { blogs, categories, categorySlug } = props;
  const blogCategoriesListProps = useBlogCategoryList();

  const cardImage = {
    src: "/images/illustrations/teacher-carrying-stuff-237-286.png",
    alt: "",
  };

  const blogListItems = blogs.map(blogToBlogListItem);
  const theme = useTheme();
  const HEADER_HEIGHT = theme.header.height;

  return (
    <Layout
      seoProps={getSeoProps({
        title: "Latest Blogs & Insights",
        description:
          "Keep up to date with our latest blog posts, filled with insights, news and updates from Oak National Academy.",
      })}
      $background="white"
      breadcrumbs={getBlogBreadcrumbs(categories, categorySlug)}
    >
      <MobileBlogFilters
        categoryListProps={{
          categories,
          selectedCategorySlug: categorySlug,
        }}
      />
      <MaxWidth $pt={[0, 80, 80]}>
        <SummaryCard
          title={"Blog Listing"}
          heading={"Inspiration for inside and outside the classroom"}
          // TODO: Replace line summary with new field from CMS
          summary={
            "Read blogs from our in-house experts to find ideas to take away and try, from curriculum planning to lesson delivery. Plus, keep up to date with the latest news and insights from Oak."
          }
          imageProps={cardImage}
        />
        <Grid $ph={[12, 0]}>
          <GridArea $order={[0, 2]} $colSpan={[12, 4, 3]}>
            <Box
              $display={["none", "block"]}
              $position={[null, "sticky"]}
              $top={[null, HEADER_HEIGHT]}
              $mt={[0, 24]}
              $pt={[48]}
            >
              <Heading
                tag="h3"
                $font="body-3"
                id={blogCategoriesListProps.labelId}
              >
                Categories
              </Heading>
              <BlogCategoryList
                labelledBy={blogCategoriesListProps.labelId}
                $mt={24}
                categories={categories}
                selectedCategorySlug={categorySlug}
              />
            </Box>
          </GridArea>
          {/* @todo is there a nicer way to make this 1 column spacer? */}
          <GridArea $order={1} $colSpan={[12, 1]} />
          <GridArea $order={[1, 0]} $colSpan={[12, 7, 8]} $mt={[48, 72]}>
            <BlogList items={blogListItems} withContainingHrs withPagination />
          </GridArea>
        </Grid>
      </MaxWidth>
      <BlogListJsonLd blogs={props.blogs} />
    </Layout>
  );
};

export const blogToBlogListItem = (
  blog: SerializedBlogPostPreview
): BlogListItemProps => ({
  contentType: "blog-post",
  title: blog.title,
  href: `/blog/${blog.slug}`,
  snippet: blog.summary,
  titleTag: "h3",
  category: blog.category,
  date: blog.date,
  mainImage: blog?.mainImage,
});

export const serializeDate = <T extends { date: Date }>(
  item: T
): T & { date: string } => ({
  ...item,
  date: item.date.toISOString(),
});

export const getStaticProps: GetStaticProps<
  BlogListingPageProps,
  // @todo is below typesafe?
  { categorySlug?: string }
> = async (context) => {
  const isPreviewMode = context.preview === true;

  const blogResults = await CMSClient.blogPosts({
    previewMode: isPreviewMode,
  });

  const blogCategories = uniqBy(
    "title",
    blogResults.map((blogResult) => blogResult.category)
  ).sort((a, b) => (a.title < b.title ? -1 : 1));

  const categorySlug = context.params?.categorySlug || null;
  const blogs = blogResults.map(serializeDate).filter((blog) => {
    if (categorySlug) {
      return blog.category.slug === categorySlug;
    }
    return true;
  });

  const results: GetStaticPropsResult<BlogListingPageProps> = {
    props: {
      blogs,
      categories: blogCategories,
      categorySlug,
    },
  };
  const resultsWithIsr = decorateWithIsr(results);

  return resultsWithIsr;
};

type URLParams = { categorySlug: string };
export const getStaticPaths: GetStaticPaths<URLParams> = async () => {
  const blogResults = await CMSClient.blogPosts();

  const paths = blogResults.map((blogResult) => ({
    params: {
      categorySlug: blogResult.category.slug,
    },
  }));

  return {
    paths,
    fallback: "blocking",
  };
};

export default BlogListingPage;
