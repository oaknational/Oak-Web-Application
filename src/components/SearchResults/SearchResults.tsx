import { SearchHit } from "../../pages/search";

interface SearchResultsProps {
  hits: Array<SearchHit>;
}

const SearchResults = (props: SearchResultsProps) => {
  const { hits } = props;
  const resultElements: Array<JSX.Element> = hits.map((hit: SearchHit) => {
    const { _source } = hit;
    const { title, id, key_stage_title } = _source;
    return (
      <li key={id}>
        {title} - <i>{key_stage_title}</i>
      </li>
    );
  });

  return <ul>{resultElements}</ul>;
};

export default SearchResults;
