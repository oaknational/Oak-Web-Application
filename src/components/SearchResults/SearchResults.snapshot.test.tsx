import { render } from "@testing-library/react";

import { SearchHit } from "../../pages/search";

import SearchResults from "./SearchResults";

const searchHits: SearchHit[] = [
  {
    _source: {
      id: 2957,
      is_sensitive: false,
      is_specialist: null,
      key_stage_slug: "key-stage-4",
      key_stage_title: "Key Stage 4",
      lesson_description:
        "In this lesson, we will look at Macbeth's transformation across the play before exploring how Shakespeare presents Macbeth as a tragic hero in Act 5, Scene 3.",
      slug: "the-tragedy-of-macbeth-cdjpac",
      subject_slug: "english",
      subject_title: "English",
      theme_title: "First Teaching",
      title: "The Tragedy of Macbeth",
      topic_slug: "macbeth-a8f1",
      topic_title: "Macbeth",
      type: "lesson",
      year_slug: "year-10",
      year_title: "Year 10",
    },
  },
];

it("renders search page unchanged", async () => {
  const { container } = render(<SearchResults hits={searchHits} />);
  expect(container).toMatchSnapshot();
});
