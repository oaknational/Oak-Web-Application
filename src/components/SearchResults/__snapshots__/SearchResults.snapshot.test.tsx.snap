// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders search page unchanged 1`] = `
<div>
  <ul>
    <li>
      The Tragedy of Macbeth
       - 
      <i>
        Key Stage 4
      </i>
    </li>
  </ul>
</div>
`;
